subsys/secure_storage/src/its/store/zms.c

```diff
@@ -0,0 +1,123 @@
+/* Copyright (c) 2024 Nordic Semiconductor
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/secure_storage/its/store.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/fs/zms.h>
+#include <zephyr/storage/flash_map.h>
+#ifdef CONFIG_SECURE_STORAGE_ITS_IMPLEMENTATION_ZEPHYR
+#include <zephyr/secure_storage/its/transform.h>
+#endif
+
+LOG_MODULE_DECLARE(secure_storage, CONFIG_SECURE_STORAGE_LOG_LEVEL);
+
+BUILD_ASSERT(CONFIG_SECURE_STORAGE_ITS_STORE_ZMS_SECTOR_SIZE
+	     > 2 * CONFIG_SECURE_STORAGE_ITS_MAX_DATA_SIZE);
+
+#define PARTITION_DT_NODE DT_CHOSEN(secure_storage_its_partition)
+
+static struct zms_fs s_zms = {
+	.flash_device = FIXED_PARTITION_NODE_DEVICE(PARTITION_DT_NODE),
+	.offset = FIXED_PARTITION_NODE_OFFSET(PARTITION_DT_NODE),
+	.sector_size = CONFIG_SECURE_STORAGE_ITS_STORE_ZMS_SECTOR_SIZE,
+};
+
+static int init_zms(void)
+{
+	int ret;
+
+	s_zms.sector_count = FIXED_PARTITION_NODE_SIZE(PARTITION_DT_NODE) / s_zms.sector_size;
+	LOG_DBG("Using %u sectors of size %u.", s_zms.sector_count, s_zms.sector_size);
+
+	ret = zms_mount(&s_zms);
+	if (ret) {
+		LOG_DBG("Mounting failed. (%d)", ret);
+	}
+	return ret;
+}
+SYS_INIT(init_zms, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+
+/* Bit position of the ITS caller ID in the ZMS entry ID. */
+#define ITS_CALLER_ID_POS 30
+/* Make sure that every ITS caller ID fits in ZMS entry IDs at the defined position. */
+BUILD_ASSERT(1 << (32 - ITS_CALLER_ID_POS) >= SECURE_STORAGE_ITS_CALLER_COUNT);
+
+static bool has_forbidden_bits_set(secure_storage_its_uid_t uid)
+{
+	if (uid.uid & GENMASK64(63, ITS_CALLER_ID_POS)) {
+		LOG_DBG("UID %u/0x%llx cannot be used as it has bits set past "
+			"the first " STRINGIFY(ITS_CALLER_ID_POS) " ones.",
+			uid.caller_id, (unsigned long long)uid.uid);
+		return true;
+	}
+	return false;
+}
+
+static uint32_t zms_id_from(secure_storage_its_uid_t uid)
+{
+	return (uint32_t)uid.uid | (uid.caller_id << ITS_CALLER_ID_POS);
```

Comment:
The result obtained from this is not used before `uid` is checked to not have any upper bits set (in `has_forbidden_bits_set()`), so here `uid.uid` can be considered to be (even less than) a `uint32_t`.

Reply:
1. Please look at your code,
2. Where is it documented that this specific implementation of the ITS store module has this limitation on bits?
3. How to handle that one implementation of the ITS store module has this limitation, while another does not?
4. Why use a 64bit uid when only 30bit are supported?

Reply:
1. [subsys/secure_storage/Kconfig.its_store:24](https://github.com/zephyrproject-rtos/zephyr/pull/81235/files#diff-fb5682f6af21bb48291e3d8c7e6c7a2ba99a9601dfc2e97f4deabe459424350dR24-R25)
2. By letting the users of the API only use a well-defined range of UIDs. Will come in a follow-up PR as stated previously.
3. The 64-bit size originally comes from the specification. It could indeed be reduced when allowed values are officially defined.